third_party/nvfuser/csrc/multidevice/multigroup_fusion.cpp

```diff
@@ -0,0 +1,159 @@
+#include <fusion_segmenter.h>
+#include <ir_utils.h>
+#include <multidevice/multigroup_fusion.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+Group::Group(
+    MultiGroupFusion* multi_group_fusion,
+    bool auto_sch,
+    ProcessRankType prank,
+    c10::Device dev)
+    : SegmentedGroup((Fusion*)multi_group_fusion),
+      auto_schedule(auto_sch),
+      device(dev),
+      multi_group_fusion_(multi_group_fusion),
+      process_rank(prank) {
+  unique_id = multi_group_fusion->running_group_counter_++;
+
+  // Copy currently availabe tensors in the context
+  //  into the group's context. This includes all
+  //  global inputs and all other groups' outputs
+  //  defined so far.
+  for (auto& it : multi_group_fusion->context_tensor_map_) {
+    context_tensors.pushBack(it.first);
+  }
+}
+
+std::unique_ptr<Fusion> Group::makeFusionCopy() {
+  std::unique_ptr<Fusion> fusion_copy = std::make_unique<Fusion>();
+  // WAR: copy the complete fusion and then change the inputs and outputs.
+  // TODO: This could be implemented in a better way
+  auto original_to_copy_map =
+      Fusion::copy(multi_group_fusion_, fusion_copy.get());
+
+  // Remove original inputs
+  std::for_each(
+      fusion_copy->inputs().begin(),
+      fusion_copy->inputs().end(),
+      [&](auto input) { fusion_copy->removeInput(input); });
+  // Remove original outputs
+  std::for_each(
+      fusion_copy->outputs().begin(),
+      fusion_copy->outputs().end(),
+      [&](auto output) { fusion_copy->removeOutput(output); });
+
+  // // Add group inputs
+  std::for_each(input_vals.begin(), input_vals.end(), [&](auto input) {
+    fusion_copy->addInput(original_to_copy_map.clone(input));
+  });
+
+  // // Add group outputs
+  std::for_each(output_vals.begin(), output_vals.end(), [&](auto output) {
+    fusion_copy->addOutput(original_to_copy_map.clone(output));
+  });
+
+  return fusion_copy;
+}
+
+MultiGroupFusion::MultiGroupFusion() {
+  Fusion();
```

Comment:
I believe you could write:

```suggestion
MultiGroupFusion::MultiGroupFusion(): Fusion() {
```

Reply:
done